proposals/0001-metadata-providers.md

```diff
@@ -0,0 +1,489 @@
+
+### Impossible to implement for libraries
+
+A large portion of logs are not generated by the end-user but by libraries such as AsyncHTTPClient, Vapor etc.
+These libraries, by design, don't know about the specific metadata keys that should be included in the logs.
+Those keys are after all runtime dependent, and may change depending on what tracing system is configured by the end user.
+
+For example, a specific `Tracer` implementation would use a type representing a trace ID, and has a way of extracting
+that trace ID from a `Baggage`. But other libraries can't know about this _specific_ trace ID and therefore would not be
+able to pass such values along to their log statements.
+
+## Proposed solution
+
+To support this kind of runtime-generated metadata in `swift-log`, we need to extend the logging APIs in an open-ended way, to allow any kinds of metadata to be provided from the asynchronous context.
+
+To solve this, we propose the extension of swift-log APIs with a new concept: _metadata providers_.
+
+`MetadataProvider` is struct nested in the `Logger` type, sitting alongside `MetadataValue`
+and the `Metadata` typealias. It has a single property `metadata` which is a closure from `Baggage?` to `Metadata?`:
+
+```swift
+extension Logger {
+    public struct MetadataProvider: Sendable {
+        public var metadata: @Sendable (_ baggage: Baggage?) -> Metadata?
+
+        public init(metadata: @escaping @Sendable (Baggage?) -> Metadata?) {
+            self.metadata = metadata
+        }
+    }
+}
+```
+
+### Defining a `MetadataProvider`
+
+While `MetadataProvider`s can be created in an ad-hoc fashion, the struct may be used as a namespace
+to define providers in. A Tracing library e.g. could expose its metadata provider via a static property
+on `Logger.MetadataProvider`:
+
+```swift
+extension Logger.MetadataProvider {
+    static let tracer = Logger.MetadataProvider { baggage in
+        guard let spanContext = baggage?.spanContext else { return nil }
+        return [
+          "traceID": "\(spanContext.traceID)",
+          "spanID": "\(spanContext.spanID)",
+        ]
+    }
+}
+```
+
+### Using a `MetadataProvider`
+
+A `MetadataProvider` can be set up either globally, on a boot-strapped logging system:
+
+```swift
+LoggingSystem.bootstrap(
+    metadataProvider: .tracer,
+    StreamLogHandler.standardOutput
+)
+```
+
+or, on a specific logger, in case some contexts should extract additional information:
+
+```swift
+let logger = Logger(label: "example", metadataProvider: .init { baggage in
+    guard let operationID = baggage?.operationID else { return nil }
+    return ["extra/opID": "\(opID)"]
+})
+```
+
+> NOTE: Setting the metadata provider on the logger directly means the `LoggingSystem` metadata provider
+> is skipped (if defined), following how an explicitly passed handler `factory`
+> overrides the `LoggingSystem`s `factory`.
```

Comment:
Hm yeah that'd be an option if we want; it'd be `addMetadataProvider` rather than "`set`" I guess. That's probably better for most use-cases I suppose.